vortex-mask/src/lib.rs

```diff
@@ -357,85 +403,74 @@ impl Mask {
-    pub fn iter(&self) -> MaskIter {
-        if self.selectivity() > FILTER_SLICES_SELECTIVITY_THRESHOLD {
-            MaskIter::Slices(self.slices())
-        } else {
-            MaskIter::Indices(self.indices())
+        match &self {
+            Self::AllTrue(len) => (*len > 0).then_some(0),
+            Self::AllFalse(_) => None,
+            Self::Values(values) => {
+                if let Some(indices) = values.indices.get() {
+                    return indices.first().copied();
+                }
+                if let Some(slices) = values.slices.get() {
+                    return slices.first().map(|(start, _)| *start);
+                }
+                values.buffer.set_indices().next()
+            }
         }
     }
 
     /// Slice the mask.
     pub fn slice(&self, offset: usize, length: usize) -> Self {
-        if self.true_count() == 0 {
-            return Self::new_false(length);
-        }
-        if self.true_count() == self.len() {
-            return Self::new_true(length);
+        assert!(offset + length <= self.len());
+        match &self {
+            Self::AllTrue(_) => Self::new_true(length),
+            Self::AllFalse(_) => Self::new_false(length),
+            Self::Values(values) => Self::from_buffer(values.buffer.slice(offset, length)),
         }
+    }
 
-        if let Some(buffer) = self.0.buffer.get() {
-            return Self::from_buffer(buffer.slice(offset, length));
+    /// Return the boolean buffer representation of the mask.
+    pub fn boolean_buffer(&self) -> AllOr<&BooleanBuffer> {
+        match &self {
+            Self::AllTrue(_) => AllOr::All,
+            Self::AllFalse(_) => AllOr::None,
+            Self::Values(values) => AllOr::Some(&values.buffer),
         }
+    }
 
-        let end = offset + length;
-
-        if let Some(indices) = self.0.indices.get() {
-            let indices = indices
-                .iter()
-                .copied()
-                .skip_while(|idx| *idx < offset)
-                .take_while(|idx| *idx < end)
-                .map(|idx| idx - offset)
-                .collect();
-            return Self::from_indices(length, indices);
+    /// Return the indices representation of the mask.
+    pub fn indices(&self) -> AllOr<&[usize]> {
+        match &self {
+            Self::AllTrue(_) => AllOr::All,
+            Self::AllFalse(_) => AllOr::None,
+            Self::Values(values) => AllOr::Some(values.indices()),
         }
+    }
 
-        if let Some(slices) = self.0.slices.get() {
-            let slices = slices
-                .iter()
-                .copied()
-                .skip_while(|(_, e)| *e <= offset)
-                .take_while(|(s, _)| *s < end)
-                .map(|(s, e)| (s.max(offset), e.min(end)))
-                .collect();
-            return Self::from_slices_unchecked(length, slices);
+    /// Return the slices representation of the mask.
+    pub fn slices(&self) -> AllOr<&[(usize, usize)]> {
```

Comment:
I'm not sure an AllOr makes sense here when the AllTrue and AllFalse cases have very obvious definitions of &[Range]

Reply:
The definition exists, but I have no way of returning one without some heap allocation somewhere.